internal/msgstream/mq_msgstream.go

```diff
@@ -510,6 +551,54 @@ func (ms *mqMsgStream) Chan() <-chan *MsgPack {
 	return ms.receiveBuf
 }
 
+func (ms *mqMsgStream) SeekReaders(msgPositions []*internalpb.MsgPosition) error {
+	for _, mp := range msgPositions {
+		reader, ok := ms.readers[mp.ChannelName]
+		if !ok {
+			return fmt.Errorf("channel %s not subscribed", mp.ChannelName)
+		}
+		messageID, err := ms.client.BytesToMsgID(mp.MsgID)
+		if err != nil {
+			return err
+		}
+		log.Debug("MsgStream reader begin to seek", zap.Any("MessageID", mp.MsgID))
+		err = reader.Seek(messageID)
+		if err != nil {
+			log.Debug("Failed to seek", zap.Error(err))
+			return err
+		}
+	}
+	return nil
+}
+
+func (ms *mqMsgStream) Next(ctx context.Context, channelName string) (TsMsg, error) {
+	reader, ok := ms.readers[channelName]
+	if !ok {
+		return nil, fmt.Errorf("reader for channel %s is not exist", channelName)
+	}
+	if reader.HasNext() {
+		msg, err := reader.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+		tsMsg, err := ms.getTsMsgFromConsumerMsg(msg)
+		if err != nil {
+			log.Error("Failed to getTsMsgFromConsumerMsg", zap.Error(err))
+			return nil, errors.New("Failed to getTsMsgFromConsumerMsg")
+		}
+		pos := tsMsg.Position()
+		tsMsg.SetPosition(&MsgPosition{
+			ChannelName: pos.ChannelName,
+			MsgID:       pos.MsgID,
+			Timestamp:   tsMsg.BeginTs(),
+		})
+		return tsMsg, nil
+	}
+	log.Debug("All data has been read, there is no more data", zap.String("channel", channelName))
+	return nil, nil
```

Comment:
Feels weird that we combine `Next` & `HasNext` next logic into one.

Reply:
Because pulsar **Next** is a blocking interface, this position may be blocked in the **Next** function if the hasNext judgment is not used.

Reply:
True. But it's still weird that this function returns `nil, nil` when there is no msg to read. Maybe we should either provide the HasNext() interface or returns `nil, NoNextError`. Since the implication of returning nil error is that the msg returns should be valid.